draft-ietf-jsonpath-base.md

```diff
@@ -1521,19 +1521,27 @@ A function expression is well-typed if all of the following are true:
   to have a result type of `LogicalType`, `SingleNodeType`, or `NodesType`.
 * If it occurs directly as a `comparable` in a comparison, the
   function is declared to have a result type of `ValueType` or `SingleNodeType`.
-* Otherwise, it occurs as an argument to a further function
-  expression and the following rules for function arguments apply to
-  its declared result type.
 * Each argument of the function conforms to the declared type of the corresponding declared
-  parameter according to one of the following rules:
-   * The argument is a function expression with declared result type that is the same as
-     the declared type of the parameter.
-   * The argument is a function expression with declared result type `SingleNodeType` and
-     the declared type of the parameter is `NodesType`.
-   * The argument is a literal primitive value and the declared type of the parameter is `ValueType`.
-   * The argument is a Singular Path and the declared type of the parameter is `SingleNodeType`.
-   * The argument is a Singular Path or `filter-path` (which includes
-     Singular Paths) and the declared type of the parameter is `NodesType`.
+  parameter according to one of the following scenarios:
+   * The declared type of the parameter is `ValueType` and the argument is one of:
+     * a function with declared type of `ValueType` or `SingleNodeType`
+     * a `singular-path`
+     * a `literal`
+   * The declared type of the parameter is `LogicalType` and the argument is one of:
+     * a function with declared type of `LogicalType`, `SingleNodeType`, or `NodesType`
+     * a `filter-path` (which includes `singular-path`)
+   * The declared type of the parameter is `SingleNodeType` and the argument is one of:
+     * a function with declared type of `SingleNodeType`
+     * a `singular-path`
+   * The declared type of the parameter is `NodesType` and the argument is one of:
+     * a function with declared type of `SingleNodeType` or `NodesType`
+     * a `filter-path`
+
+### Processing Nodelists as Arguments
+
+When any `singular-path` or function expression with declared type of `SingleNodeType` appears as an argument for a parameter of type `ValueType`, each such path or function expression is replaced by the value of its node or `Nothing` if it is empty.
```

Comment:
There's an editorial wrinkle in that `Nothing` isn't really empty. How about the following wordier, but hopefully clearer possibility? If you don't like this, I'd be happy to apply it as an editorial tweak if this PR is merged.
```suggestion
When any `singular-path` appears as an argument for a parameter of type `ValueType`, the path is replaced by the value of the node in the resulting nodelist or `Nothing` if the resulting nodelist is empty.

When any function expression with declared type of `SingleNodeType` appears as an argument for a parameter of type `ValueType`, the function expression is replaced by the value of the node in the resulting nodelist or `Nothing` if the function expression returns `Nothing`.
```

Reply:
Maybe this tongue-in-cheek kebaberizing will help:

> When any `singular-path` or function-expression-with-declared-type-of-`SingleNodeType` appears as an argument for a parameter of type `ValueType`, each such path or function expression is replaced by the value of its node or `Nothing` if it is empty.

I agree that it's a poor result of English grammar.  Maybe if I reverse them:

> When any function expression with declared type of `SingleNodeType` or `singular-path` appears as an argument for a parameter of type `ValueType`, each such path or function expression is replaced by the value of its node or `Nothing` if it is empty.

It still has the grammatical issue, but since a function can't result in a `singular-path`, maybe it makes more sense.

Reply:
Let's not waste more time on editorial niceties then.